don't annotate StopLoss variants that are immediately followed by another stop codon

iskandr opened the pull request:
Fixes https://github.com/hammerlab/varcode/issues/56

Also:
* Adds "warmup" to timing benchmarks 
* Make sure PrematureStop is actually generating a truncated protein (some variants that end in Stop codons inserted at the end may actually extend the protein)

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/varcode/57)
<!-- Reviewable:end -->

tavinathanson commented:
How about adding at least 1 test that would fail without the fix?

---

Reviewed files:
- test/test_timings.py @ r1
- varcode/effects.py @ r1
- varcode/frameshift_coding_effect.py @ r1
- varcode/in_frame_coding_effect.py @ r1
- varcode/maf.py @ r1
- varcode/variant.py @ r1

---

<sup>**[varcode/effects.py, line 493 \[r1\]](https://reviewable.io:443/reviews/hammerlab/varcode/57#-Jmsj4M5XPyWHAkNR-pB)** ([raw file](https://github.com/hammerlab/varcode/blob/963d0f7ad8e60e73ad412a0b3b9ec76d92764849/varcode/effects.py#L493)):</sup>
Weird, Reviewable doesn't recognize this without adjusting the settings to not collapse whitespace changes.

---

<sup>**[varcode/frameshift_coding_effect.py, line 73 \[r1\]](https://reviewable.io:443/reviews/hammerlab/varcode/57#-JmsltA8Q4fm-7rU_Y4y)** ([raw file](https://github.com/hammerlab/varcode/blob/963d0f7ad8e60e73ad412a0b3b9ec76d92764849/varcode/frameshift_coding_effect.py#L73)):</sup>
Can you clarify here that `n_skip` is the count of those same amino acids? (Or rename to something that automatically clarifies it?)

---

<sup>**[varcode/frameshift_coding_effect.py, line 96 \[r1\]](https://reviewable.io:443/reviews/hammerlab/varcode/57#-JmsjiXAMFFcIywFafCA)** ([raw file](https://github.com/hammerlab/varcode/blob/963d0f7ad8e60e73ad412a0b3b9ec76d92764849/varcode/frameshift_coding_effect.py#L96)):</sup>
It took me a few minutes to understand why this was a `StopLoss`, and to understand the `n_skip` stuff.

I think a bit more commenting could be helpful. This was helpful for me to write out for myself: `n_skip` = # of amino acids that are the same as the original protein. When all the amino acids are the same as the original, we either have the original protein or we've extended it. If we've extended it, it means we must have lost our stop codon.

(That last sentence, while obvious in hindsight, wasn't immediately obvious to me.)

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/varcode/57)
<!-- Sent from Reviewable.io -->

iskandr commented:
Added the variant from your issue to `test_problematic_variants`

---





<sup>**[varcode/effects.py, line 493 \[r1\]](https://reviewable.io:443/reviews/hammerlab/varcode/57#-Jmsj4M5XPyWHAkNR-pB)** ([raw file](https://github.com/hammerlab/varcode/blob/963d0f7ad8e60e73ad412a0b3b9ec76d92764849/varcode/effects.py#L493)):</sup>
Huh, that's odd

---

<sup>**[varcode/frameshift_coding_effect.py, line 73 \[r1\]](https://reviewable.io:443/reviews/hammerlab/varcode/57#-JmsltA8Q4fm-7rU_Y4y)** ([raw file](https://github.com/hammerlab/varcode/blob/963d0f7ad8e60e73ad412a0b3b9ec76d92764849/varcode/frameshift_coding_effect.py#L73)):</sup>
I got rid of this loop in favor of a helper I already wrote: `string_helpers.trim_shared_prefix` and renamed some of the variables to hopefully be clearer. 

---

<sup>**[varcode/frameshift_coding_effect.py, line 96 \[r1\]](https://reviewable.io:443/reviews/hammerlab/varcode/57#-JmsjiXAMFFcIywFafCA)** ([raw file](https://github.com/hammerlab/varcode/blob/963d0f7ad8e60e73ad412a0b3b9ec76d92764849/varcode/frameshift_coding_effect.py#L96)):</sup>
That last sentence makes for a good comment :-)

---


---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/varcode/57)
<!-- Sent from Reviewable.io -->

tavinathanson commented:
Merge away!

---

Reviewed files:
- test/test_problematic_variants.py @ r2
- varcode/frameshift_coding_effect.py @ r2

---





Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/varcode/57)
<!-- Sent from Reviewable.io -->
